### DISCUSSION TITLE
Netbox & LiberNMS Intergration

### DISCUSSION THREAD

**Discussion by vinnie12** [netbox-community/netbox#10395] [Ideas]
Hi Guys 

I am net to network automation and would like some direction on home to integrate Netbox with LibreNMS being the source for the information that is provided to Netbox.

I have also tired creating Scripts such as the NewSiteScript however it does not seem to work....

Your assistance is highly Appreciated !!

Thanks in advance...

**Comment by vinnie12** [comment]
![image](https://user-images.githubusercontent.com/37549044/190828775-d29bd83c-94ec-47cc-ab9a-c34356fffebe.png)

**Reply by vinnie12** [reply]
This is the error message for the NewSite Script

**Comment by vinnie12** [comment]
![image](https://user-images.githubusercontent.com/37549044/190853107-01350a87-5bf7-4870-8777-540ad54bca96.png)

**Comment by vinnie12** [comment]
It continues to say Results Pending 

Any suggestions??

**Comment by candlerb** [comment]
> I am net to network automation and would like some direction on home to integrate Netbox with LibreNMS being the source for the information that is provided to Netbox.

I think you are going about this the wrong way.

Netbox is not a *portal* on collected monitoring information - in other words, it its not intended to collate or "pull" data in. Netbox is a *source of truth*, which can be used to "push" instructions to other systems and control automated processes.

For more information, please read:
https://docs.netbox.dev/en/stable/introduction/#design-philosophy
https://docs.netbox.dev/en/stable/introduction/#what-netbox-is-not
and other discussion threads e.g. #10238, #9618, #7303.

With this view in mind, the right approach would be for Netbox to control LibreNMS (so when you add a device to Netbox, it automatically gets added to LibreNMS), not the other way round.  You'd have to build that integration yourself, unless someone has already done it.

Personally I use Prometheus for monitoring, and it collects the list of targets to monitor using [netbox-plugin-prometheus-sd](https://github.com/FlxPeters/netbox-plugin-prometheus-sd).

> I have also tired creating Scripts such as the NewSiteScript however it does not seem to work....

Which exact script, and which version of Netbox are you running?

The Netbox models and internal APIs do change between versions, I suspect that the script you've found isn't compatible with the version of Netbox you're running.

If the job stays in "Pending" indefinitely, make sure the "netbox-rq" worker process is running.

**Reply by vinnie12** [reply]
Thanks for the feedback 

I am currently using (v3.3.3-dev) 

We have a large environment with Cisco products and VLANS and BGP, VRF configs which currently can be viewed on Libre so i was looking an quick way to update Netbox with all the information..


I am using the NewBranchScript which i have added below.

**Comment by vinnie12** [comment]
```python
from django.utils.text import slugify

from dcim.choices import DeviceStatusChoices, SiteStatusChoices
from dcim.models import Device, DeviceRole, DeviceType, Manufacturer, Site
from extras.scripts import *


class NewBranchScript(Script):

    class Meta:
        name = "New Branch"
        description = "Provision a new branch site"
        field_order = ['switch_model', 'manufacturer', 'switch_count', 'site_name']

    site_name = StringVar(
        description="Name of the new site"
    )
    switch_count = IntegerVar(
        description="Number of access switches to create"
    )
    manufacturer = ObjectVar(
        model=Manufacturer,
        required=False
    )
    switch_model = ObjectVar(
        description="Access switch model",
        model=DeviceType,
        query_params={
            'manufacturer_id': '$manufacturer'
        }
    )

    def run(self, data, commit):

        # Create the new site
        site = Site(
            name=data['site_name'],
            slug=slugify(data['site_name']),
            status=SiteStatusChoices.STATUS_PLANNED
        )
        site.save()
        self.log_success(f"Created new site: {site}")

        # Create access switches
        switch_role = DeviceRole.objects.get(name='Access Switch')
        for i in range(1, data['switch_count'] + 1):
            switch = Device(
                device_type=data['switch_model'],
                name=f'{site.slug}-switch{i}',
                site=site,
                status=DeviceStatusChoices.STATUS_PLANNED,
                device_role=switch_role
            )
            switch.save()
            self.log_success(f"Created new switch: {switch}")

        # Generate a CSV table of new devices
        output = [
            'name,make,model'
        ]
        for switch in Device.objects.filter(site=site):
            attrs = [
                switch.name,
                switch.device_type.manufacturer.name,
                switch.device_type.model
            ]
            output.append(','.join(attrs))

        return '\n'.join(output)
```

**Reply by candlerb** [reply]
You didn't say where you got this script from. If you want help debugging it, you should talk to the script author.

Incidentally, you should never use the "develop" branch of Netbox, unless you're developing Netbox itself.  Upgrades from one version of the develop branch to another are not supported; nor are upgrades from the develop branch to master.  Basically, you could be required to throw away your entire database and start again, at any point in the future.  If you're installing Netbox from git, use the "master" branch.

**Comment by vinnie12** [comment]
I am new to this 

so i have installed " pip install netbox-plugin-prometheus-sd" 
and edited the config file 

Is there anything additional that i need to do 

Cause i do not see the the plugin

**Reply by candlerb** [reply]
Sorry: I wasn't suggesting that you should change from LibreNMS to Prometheus.  Prometheus is wonderful but it has a steep learning curve; installing the plugin is the least of it.  If you're comfortable with LibreNMS, then stick with that.

The point I was trying to make is that data should flow *from* Netbox to whatever system(s) you're managing - and I was just using my Prometheus setup as an example.

A one-time import of data collected from other systems is a reasonable thing to do.  You can use CSV imports for that, or the REST API.  I just don't think you should be trying to sync data from LibreNMS to Netbox on an ongoing basis: that goes completely against the design philosophy of Netbox.  If you want to look at data collected by LibreNMS, then use LibreNMS's own web interface for that.

**Reply by vinnie12** [reply]
Thanks for the advise
